Add Spawner.delete_forever

nsshah1288 opened the pull request:
related to this PR: https://github.com/jupyterhub/kubespawner/pull/475

tested with sqlite pvc

minrk commented:
Thanks for the PR! Unfortunately, it's not appropriate to have any kubernetes-specific methods in this repo like references to PVCs. Spawners having a hook to cleanup resources on user-delete is an interesting challenge because there may not be any existing Spawners at the time the user is deleted, but we can explore it.

Maybe a generic `delete_forever()` hook method to be called when the user is to be deleted that should be called by jupyterhub?

nsshah1288 commented:
Hi @minrk , thanks for the reply!

I have changed the code a bit, what do you think? I moved the kubernetes-specific part to spawner.py here: https://github.com/jupyterhub/kubespawner/pull/475/files

Now would it be like all other Spawners would need this method, `delete_forever` as part of their spawner base class?

Also, could you please explain more on this part? "there may not be any existing Spawners at the time the user is deleted" -- not sure I understand this

I tested this change again with sqlite pvc, by using the admin console.
Thanks!

minrk commented on jupyterhub/apihandlers/users.py:
```diff
@@ -237,6 +237,12 @@ async def delete(self, name):
+        # Delete the user's storage, if it exists.
+        # In kubespawner spawner.py, we have delete_forever
+        # method to delete the user's PVC.
+        user.spawner.delete_forever()
```
This should match the `await maybe_future()` above to allow async operations. No need for kubespawner discussion in the comments. If there's a comment, use "allow the spawner to cleanup any persistent resources associated with the user.

Probably even want try/except here because I suspect we don't want a spawner operation error to prevent user deletion.

nsshah1288 replied:
added 'await' and try/except, also changed comment

minrk commented on jupyterhub/tests/mocking.py:
```diff
@@ -93,6 +93,9 @@ def user_env(self, env):
     def _cmd_default(self):
         return [sys.executable, '-m', 'jupyterhub.tests.mocksu']
 
+    def delete_forever(self):
```
Add a docstring explaining what this is an when it is called:

> Called when a user is deleted. This can do things like request removal of resources such as persistent storage. Only called on stopped spawners, and is likely the last action ever taken for the user. Will only be called once on the user's default Spawner.

nsshah1288 replied:
added in https://github.com/jupyterhub/kubespawner/pull/475/files. Hope that's where you wanted it

nsshah1288 replied:
also added a docstring here in mocking.py

minrk commented on jupyterhub/tests/mocking.py:
```diff
@@ -93,6 +93,9 @@ def user_env(self, env):
         return [sys.executable, '-m', 'jupyterhub.tests.mocksu']
 
+    def delete_forever(self):
+        return "mocking_this_method"
```
Use `pass` for empty base methods:

```suggestion
        pass
```

nsshah1288 replied:
done

nsshah1288 commented:
Hi @minrk , I have tried to resolve all your comments, can you please let me know if anything else is required?

Note: This solution only works for dynamic singleuser storage, not for static. I tested with static storage as well (EFS) and I was still able to delete a user thru the admin console fine, but the underlying storage was not deleted.

minrk commented:
Great, thanks! I think this is enough fromt eh juptyerhub side.

nsshah1288 commented:
Awesome, thanks @minrk ! If this kubespawner PR also looks good, could you approve it as well? https://github.com/jupyterhub/kubespawner/pull/475/files

meeseeksmachine commented:
This pull request has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/should-remove-option-be-documented-with-delete-users-name-server/8457/2
